.github/workflows/my-ci-pipeline.yml

```diff
@@ -2,9 +2,30 @@ name: Continuous Integration
 on: [push] 
 jobs: 
  build: 
-  name: Build and test 
+  name: build
   runs-on: ubuntu-latest 
   steps: 
   - uses: actions/checkout@v2 
   - run: docker build --target test --tag todo-app:test .
-  - run: docker run --env-file ./.env.test todo-app:test
\ No newline at end of file
+  - run: docker run --env-file ./.env.test todo-app:test
+ deploy:
+  name: deploy
+  needs: build
+  runs-on: ubuntu-latest
+  if: github.event_name == 'push' && github.ref == 'refs/heads/Module8'
+  steps:
+   - uses: actions/checkout@v2
+   - run: docker login -u keithleverton -p ${{secrets.DOCKERPASSWORD}}
+   - run: docker build --target production --tag keithleverton/todo-app:$GITHUB_SHA .
+   - run: docker push keithleverton/todo-app:$GITHUB_SHA
+   - run: curl https://cli-assets.heroku.com/install-ubuntu.sh | sh
+   - name: build and push
+     env:
+      HEROKU_API_KEY: ${{secrets.HEROKU_API_KEY}} 
+     run: | 
+      heroku container:login 
+      docker tag keithleverton/todo-app:$GITHUB_SHA registry.heroku.com/to-do-app-kl/web 
+      docker push registry.heroku.com/to-do-app-kl/web 
+      heroku container:release web -a to-do-app-kl
```

Comment:
Personally I'm not a huge fan of combining many commands - if this errors it requires quite a bit more thinking to track down why that if each individual step was a single part of this. If you do take the approach of splitting it up, it might make sense to move the `env` block to the job level (e.g. so it looks like below) rather than repeating that setting where required

```
 deploy:
  name: deploy
  needs: build
  runs-on: ubuntu-latest
  if: github.event_name == 'push' && github.ref == 'refs/heads/Module8'
  env:
    HEROKU_API_KEY: ${{secrets.HEROKU_API_KEY}}
  steps:
    ...
```